File: minigit-5/ts/src/commands/commit.ts
// commit command - Create a new commit

import * as fs from 'fs';
import * as path from 'path';
import { findRepoRoot, formatAuthorDate, getAuthorInfo, getCommitterInfo, normalizePath, getFileMode } from '../utils';
import { createCommit, createTree, getTreeFromTreeIsh, walkTree, createBlob } from '../objects';
import { readIndex, writeIndex, buildTreeFromIndex } from '../index-file';
import { getHeadCommit, getCurrentBranch, updateRef, setHead, isDetachedHead } from '../refs';
import { TreeEntry, IndexEntry } from '../types';

export function commit(args: string[]): number {
  const repoRoot = findRepoRoot();
  if (!repoRoot) {
    console.error('fatal: not a minigit repository');
    return 1;
  }

  // Parse arguments
  let message: string | null = null;
  let amend = false;
  let autoStage = false;

  for (let i = 0; i < args.length; i++) {
    const arg = args[i];
    if (arg === '-m' && i + 1 < args.length) {
      message = args[i + 1];
      i++;
    } else if (arg === '--amend') {
      amend = true;
    } else if (arg === '-a') {
      autoStage = true;
    }
  }

  if (!message && !amend) {
    console.error('error: must provide commit message with -m');
    return 1;
  }

  // Auto-stage modified tracked files if -a flag
  if (autoStage) {
    const headSha = getHeadCommit(repoRoot);
    if (headSha) {
      try {
        const treeSha = getTreeFromTreeIsh(headSha, repoRoot);
        const trackedFiles = walkTree(treeSha, '', repoRoot);
        const index = readIndex(repoRoot);

        for (const [filePath] of trackedFiles) {
          const fullPath = path.join(repoRoot, filePath);
          if (fs.existsSync(fullPath)) {
            const stats = fs.lstatSync(fullPath);
            if (!stats.isDirectory()) {
              const content = stats.isSymbolicLink()
                ? Buffer.from(fs.readlinkSync(fullPath))
                : fs.readFileSync(fullPath);
              const sha = createBlob(content, true, repoRoot);
              const mode = getFileMode(fullPath);

              const existingIdx = index.entries.findIndex(e => e.path === filePath);
              const entry: IndexEntry = {
                ctimeSec: Math.floor(stats.ctimeMs / 1000),
                ctimeNsec: Math.floor((stats.ctimeMs % 1000) * 1000000),
                mtimeSec: Math.floor(stats.mtimeMs / 1000),
                mtimeNsec: Math.floor((stats.mtimeMs % 1000) * 1000000),
                dev: stats.dev,
                ino: stats.ino,
                mode,
                uid: stats.uid,
                gid: stats.gid,
                size: stats.size,
                sha,
                flags: Math.min(filePath.length, 0xfff),
                path: filePath,
              };

              if (existingIdx >= 0) {
                index.entries[existingIdx] = entry;
              } else {
                index.entries.push(entry);
              }
            }
          }
        }

        writeIndex(index, repoRoot);
      } catch {
        // No HEAD commit yet
      }
    }
  }

  // Read current index
  const index = readIndex(repoRoot);

  if (index.entries.length === 0 && !amend) {
    console.error('nothing to commit');
    return 1;
  }

  // Get current HEAD
  const headSha = getHeadCommit(repoRoot);

  // For amend, get the previous commit's message if no new message
  let parents: string[] = [];
  if (amend) {
    if (!headSha) {
      console.error('fatal: no commit to amend');
      return 1;
    }
    const { getCommit } = require('../objects');
    const prevCommit = getCommit(headSha, repoRoot);
    parents = prevCommit.parents;
    if (!message) {
      message = prevCommit.message;
    }
  } else {
    if (headSha) {
      parents = [headSha];
    }
  }

  // Check if there are any changes to commit
  if (headSha && !amend) {
    const headTreeSha = getTreeFromTreeIsh(headSha, repoRoot);
    const headFiles = walkTree(headTreeSha, '', repoRoot);

    // Compare with index
    let hasChanges = false;
    const indexFiles = new Map<string, string>();
    for (const entry of index.entries) {
      indexFiles.set(entry.path, entry.sha);
    }

    // Check for differences
    if (indexFiles.size !== headFiles.size) {
      hasChanges = true;
    } else {
      for (const [filePath, sha] of indexFiles) {
        const headEntry = headFiles.get(filePath);
        if (!headEntry || headEntry.sha !== sha) {
          hasChanges = true;
          break;
        }
      }
    }

    if (!hasChanges) {
      console.error('nothing to commit, working tree clean');
      return 1;
    }
  }

  // Build tree from index
  const treeSha = buildTreeFromIndex(repoRoot);

  // Get author and committer info
  const authorInfo = getAuthorInfo();
  const committerInfo = getCommitterInfo();

  const author = formatAuthorDate(authorInfo.name, authorInfo.email, authorInfo.date, authorInfo.tz);
  const committer = formatAuthorDate(committerInfo.name, committerInfo.email, committerInfo.date, committerInfo.tz);

  // Create commit
  const commitSha = createCommit(treeSha, parents, author, committer, message!, repoRoot);

  // Update HEAD/branch
  const branch = getCurrentBranch(repoRoot);
  if (branch) {
    updateRef(`refs/heads/${branch}`, commitSha, repoRoot);
  } else {
    // Detached HEAD
    setHead(commitSha, repoRoot);
  }

  // Output result
  const shortSha = commitSha.slice(0, 7);
  const isRoot = parents.length === 0;
  const branchInfo = branch ? ` (${branch})` : '';
  console.log(`[${branch || 'detached HEAD'} ${isRoot ? '(root-commit) ' : ''}${shortSha}] ${message!.split('\n')[0]}`);

  return 0;
}

File: minigit-5/py/commands/commit.py
"""commit command - Create a new commit."""

import sys
import os
from pathlib import Path
from datetime import datetime, timezone

sys.path.insert(0, str(Path(__file__).parent.parent))

from utils import find_repo_root, get_file_mode
from objects import Blob, Tree, TreeEntry, Commit
from index import Index, create_entry_from_file
from refs import resolve_head, update_head, get_head_ref, write_ref


def run(args: list[str]) -> int:
    """Create a new commit."""
    repo_root = find_repo_root()
    if repo_root is None:
        print("fatal: not a minigit repository", file=sys.stderr)
        return 1

    # Parse arguments
    message = None
    amend = False
    auto_stage = False

    i = 0
    while i < len(args):
        if args[i] == '-m' and i + 1 < len(args):
            message = args[i + 1]
            i += 2
        elif args[i] == '--amend':
            amend = True
            i += 1
        elif args[i] in ('-a', '--all'):
            auto_stage = True
            i += 1
        else:
            i += 1

    if message is None:
        print("fatal: commit message required (-m)", file=sys.stderr)
        return 1

    index = Index.read(repo_root)

    # Auto-stage modified tracked files if -a flag
    if auto_stage:
        auto_stage_modified(repo_root, index)
        index.write(repo_root)

    # Check if there are staged changes
    current_head = resolve_head(repo_root)

    if amend and current_head is None:
        print("fatal: nothing to amend", file=sys.stderr)
        return 1

    if not index.entries:
        print("fatal: nothing to commit", file=sys.stderr)
        return 1

    # Check if there are actual changes
    if current_head and not amend:
        current_commit = Commit.read(current_head, repo_root)
        new_tree_sha = build_tree(repo_root, index)
        if new_tree_sha == current_commit.tree_sha:
            print("nothing to commit, working tree clean", file=sys.stderr)
            return 1

    # Build tree from index
    tree_sha = build_tree(repo_root, index)

    # Get parent(s)
    parents = []
    if amend:
        old_commit = Commit.read(current_head, repo_root)
        parents = old_commit.parents
    elif current_head:
        parents = [current_head]

    # Get author/committer info
    author = os.environ.get('GIT_AUTHOR_NAME', 'Unknown')
    author_email = os.environ.get('GIT_AUTHOR_EMAIL', 'unknown@example.com')
    committer = os.environ.get('GIT_COMMITTER_NAME', author)
    committer_email = os.environ.get('GIT_COMMITTER_EMAIL', author_email)

    # Get timestamps
    author_date = format_timestamp(os.environ.get('GIT_AUTHOR_DATE'))
    committer_date = format_timestamp(os.environ.get('GIT_COMMITTER_DATE'))

    # Create commit
    commit = Commit(
        tree_sha=tree_sha,
        parents=parents,
        author=author,
        author_email=author_email,
        author_date=author_date,
        committer=committer,
        committer_email=committer_email,
        committer_date=committer_date,
        message=message
    )
    commit_sha = commit.write(repo_root)

    # Update HEAD/branch
    head_ref = get_head_ref(repo_root)
    if head_ref:
        write_ref(repo_root, head_ref, commit_sha)
    else:
        # Detached HEAD
        from refs import write_head
        write_head(repo_root, commit_sha)

    print(f"[{get_branch_name(repo_root)} {commit_sha[:7]}] {message.split(chr(10))[0]}")
    return 0


def build_tree(repo_root: Path, index: Index) -> str:
    """Build tree object(s) from index entries."""
    entries = index.get_all_entries()

    # Group entries by top-level directory
    root_entries = []
    subdirs = {}

    for entry in entries:
        parts = entry.name.split('/')
        if len(parts) == 1:
            # File at root
            root_entries.append(TreeEntry(entry.mode, entry.name, entry.sha))
        else:
            # File in subdirectory
            top_dir = parts[0]
            rest = '/'.join(parts[1:])
            if top_dir not in subdirs:
                subdirs[top_dir] = []
            subdirs[top_dir].append((rest, entry))

    # Recursively build subtrees
    for dir_name, sub_entries in subdirs.items():
        subtree_sha = build_subtree(repo_root, sub_entries)
        root_entries.append(TreeEntry(0o40000, dir_name, subtree_sha))

    tree = Tree(root_entries)
    return tree.write(repo_root)


def build_subtree(repo_root: Path, entries: list) -> str:
    """Recursively build a subtree."""
    root_entries = []
    subdirs = {}

    for rel_path, entry in entries:
        parts = rel_path.split('/')
        if len(parts) == 1:
            root_entries.append(TreeEntry(entry.mode, rel_path, entry.sha))
        else:
            top_dir = parts[0]
            rest = '/'.join(parts[1:])
            if top_dir not in subdirs:
                subdirs[top_dir] = []
            subdirs[top_dir].append((rest, entry))

    for dir_name, sub_entries in subdirs.items():
        subtree_sha = build_subtree(repo_root, sub_entries)
        root_entries.append(TreeEntry(0o40000, dir_name, subtree_sha))

    tree = Tree(root_entries)
    return tree.write(repo_root)


def format_timestamp(date_str: str = None) -> str:
    """Format a timestamp for git commit."""
    if date_str:
        # Parse ISO format date
        try:
            if date_str.endswith('Z'):
                date_str = date_str[:-1] + '+00:00'
            dt = datetime.fromisoformat(date_str)
            ts = int(dt.timestamp())
            offset = dt.strftime('%z')
            if offset:
                offset = offset[:3] + offset[3:]
            else:
                offset = '+0000'
            return f"{ts} {offset}"
        except:
            pass

    # Default to current time
    now = datetime.now(timezone.utc)
    ts = int(now.timestamp())
    return f"{ts} +0000"


def get_branch_name(repo_root: Path) -> str:
    """Get current branch name for commit message."""
    from refs import get_current_branch
    branch = get_current_branch(repo_root)
    return branch or "HEAD"


def auto_stage_modified(repo_root: Path, index: Index):
    """Auto-stage modified tracked files."""
    for name, entry in list(index.entries.items()):
        file_path = repo_root / name
        if file_path.exists():
            # Check if file is modified
            if file_path.is_symlink():
                content = os.readlink(file_path).encode()
                mode = 0o120000
            else:
                content = file_path.read_bytes()
                mode = get_file_mode(file_path)

            blob = Blob(content)
            if blob.sha != entry.sha:
                blob.write(repo_root)
                new_entry = create_entry_from_file(file_path, name, blob.sha, mode)
                index.add_entry(new_entry)
